### ISSUE TITLE
[JENKINS-37491] Build Origin PRs (merge with base branch) conducts rebuilds when baseline changes

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/github-branch-source-plugin#991]
<p>When 'Build Origin PRs (merge with base branch)' is selected any PR's which are filed against a base branch are re-triggered when the base branch changes.</p>

<p>While this is all fine if you have a small # of PR's & the builds are quick, if you have a large team & backlog of PR's then any merge to the base branch will automatically trigger every PR which is using the same baseline.</p>

<p>When you then end up with is a Jenkins that has a massive queue to process & github often complaining that you've exceeded the API rate limit.</p>

<p>The concept of building the merge branch is great, however at the moment I really can't use it because of the previously stated behaviour & causes massive issues for our builds.</p>

<p>Previous versions of the plugin didn't exhibit this behaviour at all, nor do other systems like travis which build PR's using the merge branch, nor did the github pull request builder plugin.</p>

<p>As a workaround I'm forced to use the 'Build Origin PRs (unmerged HEAD)' and then conduct the merge or checkout of the GH merge branch itself within the pipeline which is less than ideal.</p>

<p>Can we have an option to inhibit a re-trigger when the baseline changes for PRs with merge?</p>

<p>btw I like where this plugin is going</p>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/nullify005.png" title="nullify005's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=nullify005">nullify005</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37491" target="_blank">Build Origin PRs (merge with base branch) conducts rebuilds when baseline changes</a></i></summary>
<i><ul>
<li><b>status</b>: Closed
<li><b>priority</b>: Major
<li><b>component(s)</b>: github-branch-source-plugin
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2018-03-15T21:20:53+00:00
<li><b>votes</b>: 6
<li><b>watchers</b>: 20
<li><b>imported</b>: 2025-12-02
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>When 'Build Origin PRs (merge with base branch)' is selected any PR's which are filed against a base branch are re-triggered when the base branch changes.</p>

<p>While this is all fine if you have a small # of PR's &amp; the builds are quick, if you have a large team &amp; backlog of PR's then any merge to the base branch will automatically trigger every PR which is using the same baseline.</p>

<p>When you then end up with is a Jenkins that has a massive queue to process &amp; github often complaining that you've exceeded the API rate limit.</p>

<p>The concept of building the merge branch is great, however at the moment I really can't use it because of the previously stated behaviour &amp; causes massive issues for our builds.</p>

<p>Previous versions of the plugin didn't exhibit this behaviour at all, nor do other systems like travis which build PR's using the merge branch, nor did the github pull request builder plugin.</p>

<p>As a workaround I'm forced to use the 'Build Origin PRs (unmerged HEAD)' and then conduct the merge or checkout of the GH merge branch itself within the pipeline which is less than ideal.</p>

<p>Can we have an option to inhibit a re-trigger when the baseline changes for PRs with merge?</p>

<p>btw I like where this plugin is going</p></pre>
</details>
</details>
<details><summary><i>environment</i></summary>

```
github-branch-source 1.8.1<br/>
jenkins 2.18
```
</details>
<details><summary><i>1 attachment</i></summary>

- [screenshot-1.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/33788/screenshot-1.png)
  > ![screenshot-1.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/33788/screenshot-1.png)
</details>

**Comment by jenkins-infra-bot** [comment]
- _Remote link associated with this issue:_


  - [CloudBees Internal OSS-2121](https://cloudbees.atlassian.net/browse/OSS-2121)

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37491?focusedId=267205&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-267205">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Had a long discussion with <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=michaelneale" class="user-hover" rel="michaelneale">michaelneale</a> about this. Briefly, if you are <em>not</em> rebuilding when the base branch changes, then Jenkins is not really verifying that the merge product is valid. So you need to pick some tradeoff between fidelity and performance: something in between the true merge build and the head build.</p>

<p>One option would be to merge against the latest base commit which is not newer (according to recorded commit timestamp—not necessarily related to push time) than the PR head commit. This would probably be the closest to the prior behavior of this plugin, which used the poorly documented and unreliable merge ref from GitHub, and would match the behavior of your workaround if I understand it. Another option would be to merge against new base commits, but only at limited intervals, like a day or a week—so for example if you had an old PR sitting around dormant, it would periodically get rechecked against new base versions, but not on every base branch push.</p>
</pre>
</details>

</ul>
</details>

<p>Had a long discussion with <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=michaelneale" class="user-hover" rel="michaelneale">michaelneale</a> about this. Briefly, if you are <em>not</em> rebuilding when the base branch changes, then Jenkins is not really verifying that the merge product is valid. So you need to pick some tradeoff between fidelity and performance: something in between the true merge build and the head build.</p>

<p>One option would be to merge against the latest base commit which is not newer (according to recorded commit timestamp—not necessarily related to push time) than the PR head commit. This would probably be the closest to the prior behavior of this plugin, which used the poorly documented and unreliable merge ref from GitHub, and would match the behavior of your workaround if I understand it. Another option would be to merge against new base commits, but only at limited intervals, like a day or a week—so for example if you had an old PR sitting around dormant, it would periodically get rechecked against new base versions, but not on every base branch push.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/michaelneale.png" title="michaelneale's avatar" /> michaelneale:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37491?focusedId=267280&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-267280">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Yes, once explained it makes sense. Technically current PR builders (not this one) do the wrong thing and give a false sense of the state of a validation and merge ready-ness. </p>

<p>As Jesse pointed out - this is a trade of of: How often can you tolerate a broken master: 1) sometimes 3) never. </p>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick" class="user-hover" rel="jglick">jglick</a> this seems a subtle point, having behavior like exisitng PR builders would be probably interesting, although all of this is hard to communicate without arm waving, so not sure how to present it as an option (but this is probably not the first time it has come up as a point of confusion). </p>


</pre>
</details>

</ul>
</details>

<p>Yes, once explained it makes sense. Technically current PR builders (not this one) do the wrong thing and give a false sense of the state of a validation and merge ready-ness. </p>

<p>As Jesse pointed out - this is a trade of of: How often can you tolerate a broken master: 1) sometimes 3) never. </p>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick" class="user-hover" rel="jglick">jglick</a> this seems a subtle point, having behavior like exisitng PR builders would be probably interesting, although all of this is hard to communicate without arm waving, so not sure how to present it as an option (but this is probably not the first time it has come up as a point of confusion). </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/nullify005.png" title="nullify005's avatar" /> nullify005:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37491?focusedId=267283&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-267283">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Thanks for the explanations Jesse &amp; Michael,</p>

<p>While I completely understand the intent, I think we really need to have an option to inhibit the re-trigger yet keep the ability to build merge instead of the PR's HEAD.</p>

<p>In our case:</p>

<ul>
	<li>We have a team of 12 constantly raising PR's &amp; merging them into the baseline</li>
	<li>There is an average of about 10 PR's which are in an open state</li>
	<li>Each build takes about 25 mins to perform</li>
	<li>A minimum of 10 nodes are required to execute a sub-set of tests in parallel for each build</li>
</ul>


<p>When we enabled the 'merge with base branch' functionality the following very quickly occurred:</p>

<ul>
	<li>Jenkins rapidly started queuing jobs up to something like 400</li>
	<li>Our autoscaling group backing the ECS docker nodes for each parallel job went from the usually quiet 3-5 went straight to it's cap of 20</li>
	<li>Each PR ended up having multiple builds executing all on different baselines</li>
	<li>None of the developers were able to determine whether their PR's were valid because they were constantly rebuilding</li>
	<li>Github started rejecting API requests because we were over the rate limit - I imagine because branch indexing was churning over &amp; over for every github hook execution</li>
	<li>With Github check protection turned on it was not possible to merge a PR because the commit status notification was nearly always 'building'</li>
	<li>The team basically chose to ignore the CI system ... if the CI isn't the source of truth then anarchy!</li>
</ul>


<p>Perhaps if our situation was somewhat different then it'd be perfectly acceptable to let it do rebuilds when the baseline changes, however in our case (&amp; probably many others) as you can see it actually just crushes development team productivity not to mention all of the infrastructure which is attached.</p>

<p>Pretty please, can we not have to make ugly workaround hacks in the pipeline itself because we can't filter the triggering but want the git plugin(s) to do the merge for us?</p>

</pre>
</details>

</ul>
</details>

<p>Thanks for the explanations Jesse & Michael,</p>

<p>While I completely understand the intent, I think we really need to have an option to inhibit the re-trigger yet keep the ability to build merge instead of the PR's HEAD.</p>

<p>In our case:</p>

<ul>
	<li>We have a team of 12 constantly raising PR's & merging them into the baseline</li>
	<li>There is an average of about 10 PR's which are in an open state</li>
	<li>Each build takes about 25 mins to perform</li>
	<li>A minimum of 10 nodes are required to execute a sub-set of tests in parallel for each build</li>
</ul>


<p>When we enabled the 'merge with base branch' functionality the following very quickly occurred:</p>

<ul>
	<li>Jenkins rapidly started queuing jobs up to something like 400</li>
	<li>Our autoscaling group backing the ECS docker nodes for each parallel job went from the usually quiet 3-5 went straight to it's cap of 20</li>
	<li>Each PR ended up having multiple builds executing all on different baselines</li>
	<li>None of the developers were able to determine whether their PR's were valid because they were constantly rebuilding</li>
	<li>Github started rejecting API requests because we were over the rate limit - I imagine because branch indexing was churning over & over for every github hook execution</li>
	<li>With Github check protection turned on it was not possible to merge a PR because the commit status notification was nearly always 'building'</li>
	<li>The team basically chose to ignore the CI system ... if the CI isn't the source of truth then anarchy!</li>
</ul>


<p>Perhaps if our situation was somewhat different then it'd be perfectly acceptable to let it do rebuilds when the baseline changes, however in our case (& probably many others) as you can see it actually just crushes development team productivity not to mention all of the infrastructure which is attached.</p>

<p>Pretty please, can we not have to make ugly workaround hacks in the pipeline itself because we can't filter the triggering but want the git plugin(s) to do the merge for us?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/michaelneale.png" title="michaelneale's avatar" /> michaelneale:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37491?focusedId=267284&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-267284">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=nullify005" class="user-hover" rel="nullify005">nullify005</a> I had a similar situation - many open PRs, and it would exhaust things. In that case I turned of the merge build (as it was worth it in my case) but that isn't true for everyone. I think that is the only option for now - but if you really want the validated merge, then what is needed is a new option: do not rebuild when master changes, which I <b>think</b> jesse mentioned as a possibility </p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=nullify005" class="user-hover" rel="nullify005">nullify005</a> I had a similar situation - many open PRs, and it would exhaust things. In that case I turned of the merge build (as it was worth it in my case) but that isn't true for everyone. I think that is the only option for now - but if you really want the validated merge, then what is needed is a new option: do not rebuild when master changes, which I <b>think</b> jesse mentioned as a possibility </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>kmadel:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37491?focusedId=268758&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-268758">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I may not follow correctly, but I believe one more option may be to select the 'unmerged head' for all PRs - origin and fork - and then use the GitHub protected branch feature to force branches (PRs) to be up to date with the base branch before being allowed to merge:</p>

<p> <span class="image-wrap" style=""><a id="33788_thumb" href="https://issues.jenkins.io/secure/attachment/33788/33788_screenshot-1.png" title="screenshot-1.png" file-preview-type="image" file-preview-id="33788" file-preview-title="screenshot-1.png"><img src="https://issues.jenkins.io/secure/thumbnail/33788/_thumb_33788.png" style="border: 0px solid black" role="presentation"/></a></span> </p>
</pre>
</details>

</ul>
</details>

<p>I may not follow correctly, but I believe one more option may be to select the 'unmerged head' for all PRs - origin and fork - and then use the GitHub protected branch feature to force branches (PRs) to be up to date with the base branch before being allowed to merge:</p>

<p> <span class="image-wrap" style=""><a id="33788_thumb" href="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/33788/screenshot-1.png" title="screenshot-1.png" file-preview-type="image" file-preview-id="33788" file-preview-title="screenshot-1.png"><img src="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/33788/screenshot-1.png" style="border: 0px solid black" role="presentation"/></a></span> </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/stephenconnolly.png" title="stephenconnolly's avatar" /> stephenconnolly:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37491?focusedId=280824&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-280824">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>With GitHub Branch Source 2.0.0-beta-1 (available from the <a href="https://jenkins.io/blog/2013/09/23/experimental-plugins-update-center/" class="external-link" target="_blank" rel="nofollow noopener">experimental update center</a> now or 2.0.0 (available in early January 2017)</p>

<p>When you push a change to the base branch, it should  trigger only the base branch. The PR-merge branches will get rebuilt when the indexing kicks in as indexing will detect that the merge commit is now different, but that should only happen once a day/week (depending on how often you configure indexing) so should be much less of an issue</p>
</pre>
</details>

</ul>
</details>

<p>With GitHub Branch Source 2.0.0-beta-1 (available from the <a href="https://jenkins.io/blog/2013/09/23/experimental-plugins-update-center/" class="external-link" target="_blank" rel="nofollow noopener">experimental update center</a> now or 2.0.0 (available in early January 2017)</p>

<p>When you push a change to the base branch, it should  trigger only the base branch. The PR-merge branches will get rebuilt when the indexing kicks in as indexing will detect that the merge commit is now different, but that should only happen once a day/week (depending on how often you configure indexing) so should be much less of an issue</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/michaelneale.png" title="michaelneale's avatar" /> michaelneale:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37491?focusedId=280941&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-280941">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=stephenconnolly" class="user-hover" rel="stephenconnolly">stephenconnolly</a> that actually sounds like a reasonably useful solution to that problem. Build storms are common with that turned on... so mostly I have turned that off as it just wasn't practical when you have a bunch of pull requests in progress... </p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=stephenconnolly" class="user-hover" rel="stephenconnolly">stephenconnolly</a> that actually sounds like a reasonably useful solution to that problem. Build storms are common with that turned on... so mostly I have turned that off as it just wasn't practical when you have a bunch of pull requests in progress... </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>hamiltb:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37491?focusedId=280999&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-280999">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Stephen Connolly, yes that sounds like a good solution. We are considering changing our PR build support recommendation to be: 
<em>If you want Jenkins PR build support, choose the option to build the PR at head (not merge). Any changes to your task branch and PR will result in the PR being built at head. To insure that your merge doesn’t destabilize master, choose the Git “branch protection” option on Master.</em></p>
</pre>
</details>

</ul>
</details>

<p>Stephen Connolly, yes that sounds like a good solution. We are considering changing our PR build support recommendation to be: <br/>
<em>If you want Jenkins PR build support, choose the option to build the PR at head (not merge). Any changes to your task branch and PR will result in the PR being built at head. To insure that your merge doesn’t destabilize master, choose the Git “branch protection” option on Master.</em></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>hamiltb:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37491?focusedId=281001&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-281001">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Sorry I meant to say that I liked Kurt Madel's solution (branch protection). Stephens solution could still result in build storms.</p>
</pre>
</details>

</ul>
</details>

<p>Sorry I meant to say that I liked Kurt Madel's solution (branch protection). Stephens solution could still result in build storms.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/michaelneale.png" title="michaelneale's avatar" /> michaelneale:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37491?focusedId=281029&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-281029">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>It's also hard to describe all these options to people - the reality is "do you want lots of builds when you merge to master" - but most people would say "no".</p>
</pre>
</details>

</ul>
</details>

<p>It's also hard to describe all these options to people - the reality is "do you want lots of builds when you merge to master" - but most people would say "no".</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>mapi:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37491?focusedId=282322&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-282322">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I think, the most of people(include me) want...</p>

<p><b>Build Trigger:</b>
When PR's head(unmerged) updated.</p>

<p><b>Build Target(what to checkout):</b>
merged commit with base branch.(build with merged Jenkinsfile)</p>
</pre>
</details>

</ul>
</details>

<p>I think, the most of people(include me) want...</p>

<p><b>Build Trigger:</b><br/>
When PR's head(unmerged) updated.</p>

<p><b>Build Target(what to checkout):</b><br/>
merged commit with base branch.(build with merged Jenkinsfile)</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>rainwaj:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37491?focusedId=290260&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-290260">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Is this addressed by the Github Branch Source updates?</p>
</pre>
</details>

</ul>
</details>

<p>Is this addressed by the Github Branch Source updates?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37491?focusedId=290264&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-290264">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=rainwaj" class="user-hover" rel="rainwaj">rainwaj</a>: no.</p>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=mapi" class="user-hover" rel="mapi">mapi</a> you more or less described what I already wrote:</p>

<blockquote><p>merge against the latest base commit which is not newer […] than the PR head commit</p></blockquote>

<p>under the assumption that the build starts quickly and the checkout in fact comes before any other base branch commit lands. If the checkout is delayed (for example because of a busy build queue), the behavior I described is preferable to simply merging against the tip of the base branch in that it is more deterministic what the merge commit will be.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=rainwaj" class="user-hover" rel="rainwaj">rainwaj</a>: no.</p>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=mapi" class="user-hover" rel="mapi">mapi</a> you more or less described what I already wrote:</p>

<blockquote><p>merge against the latest base commit which is not newer […] than the PR head commit</p></blockquote>

<p>under the assumption that the build starts quickly and the checkout in fact comes before any other base branch commit lands. If the checkout is delayed (for example because of a busy build queue), the behavior I described is preferable to simply merging against the tip of the base branch in that it is more deterministic what the merge commit will be.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37491?focusedId=292159&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-292159">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>By the way it seems that currently while a full reindex of the repository will trigger builds of all merge PR jobs which have a new base branch commit, an event (such as webhook) updating the base branch does not immediately trigger those builds.</p>
</pre>
</details>

</ul>
</details>

<p>By the way it seems that currently while a full reindex of the repository will trigger builds of all merge PR jobs which have a new base branch commit, an event (such as webhook) updating the base branch does not immediately trigger those builds.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>spencermalone:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37491?focusedId=292610&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-292610">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Is there a way to rate limit those builds triggered by the full reindexes? We have a large (2 gig) repo with 60-100 PRs open at any one time, and lots of PR churn. When that indexing kicks in, it queues all the PRs for a rebuild, and often things start breaking as the master tries to check out all those PRs at once. The problem is I can't seem to configure the master's checkout behavior at all (referring to the master's flyweight executors checking things out before we run the Jenkinsfile).</p>
</pre>
</details>

</ul>
</details>

<p>Is there a way to rate limit those builds triggered by the full reindexes? We have a large (2 gig) repo with 60-100 PRs open at any one time, and lots of PR churn. When that indexing kicks in, it queues all the PRs for a rebuild, and often things start breaking as the master tries to check out all those PRs at once. The problem is I can't seem to configure the master's checkout behavior at all (referring to the master's flyweight executors checking things out before we run the Jenkinsfile).</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/stephenconnolly.png" title="stephenconnolly's avatar" /> stephenconnolly:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37491?focusedId=292642&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-292642">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=hrmpw" class="user-hover" rel="hrmpw">hrmpw</a> FYI another use-case ^^^</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=hrmpw" class="user-hover" rel="hrmpw">hrmpw</a> FYI another use-case ^^^</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>rainwaj:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37491?focusedId=292655&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-292655">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>+1 to <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=spencermalone" class="user-hover" rel="spencermalone">spencermalone</a>'s comment. I was hoping there might be an additional option in recent plugin updates to only rebuild PRs when folks commit to their PR and not when the base branch is updated. </p>
</pre>
</details>

</ul>
</details>

<p>+1 to <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=spencermalone" class="user-hover" rel="spencermalone">spencermalone</a>'s comment. I was hoping there might be an additional option in recent plugin updates to only rebuild PRs when folks commit to their PR and not when the base branch is updated. </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>spencermalone:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37491?focusedId=292662&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-292662">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I'm a fan of the current behavior, it helps ensure safety on PRs that are left open for weeks. If we're going to stay that route, I just need some way to just say "Don't let the master try to checkout more than x repos at once for the purposes of changelog + Jenkinsfile loading"</p>
</pre>
</details>

</ul>
</details>

<p>I'm a fan of the current behavior, it helps ensure safety on PRs that are left open for weeks. If we're going to stay that route, I just need some way to just say "Don't let the master try to checkout more than x repos at once for the purposes of changelog + Jenkinsfile loading"</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>rainwaj:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37491?focusedId=292867&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-292867">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Related question: is there a way to differentiate whether a rebuild was triggered by update to base branch or head? I've setup our pipeline job so that I differentiate whether a job is manually or autotriggered by checking currentBuild.rawBuild.getCause(hudson.model.Cause$UserIdCause).properties.shortDescription for "Started by".</p>

<p>It would be super helpful if the "Branch Indexing" message were more granular since I may want to take different action with my build depending on the event that triggered the hook, etc. Thanks</p>
</pre>
</details>

</ul>
</details>

<p>Related question: is there a way to differentiate whether a rebuild was triggered by update to base branch or head? I've setup our pipeline job so that I differentiate whether a job is manually or autotriggered by checking currentBuild.rawBuild.getCause(hudson.model.Cause$UserIdCause).properties.shortDescription for "Started by".</p>

<p>It would be super helpful if the "Branch Indexing" message were more granular since I may want to take different action with my build depending on the event that triggered the hook, etc. Thanks</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>rainwaj:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37491?focusedId=293079&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-293079">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I wanted to add here a solution I found that feels like a nice workaround at least for my needs:</p>


<ol>
	<li>In GitHub Org options check to build both merged and unmerged PRs so you have a bunch of PR jobs ending in -merge and -head.</li>
	<li>In the auto branch triggering field add PR-\d+-head so you auto build all head jobs but not all merge jobs. This will avoid the build storm that causes so many problems.</li>
	<li>From inside the Jenkinsfile add something like this:
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java"><span class="code-comment">// have head job trigger merge job...
</span><span class="code-keyword">if</span> (autoTriggered() &amp;&amp; env.BRANCH_NAME.endsWith(<span class="code-quote">'head'</span>)) {
    def headJobName = <span class="code-quote">"${env.BRANCH_NAME}"</span>
    def mergeJobName = (headJobName.substring(0, headJobName.length() - 4)).concat(<span class="code-quote">'merge'</span>) <span class="code-comment">// replace head with merge
</span>    build job: mergeJobName, wait: <span class="code-keyword">false</span>
    <span class="code-keyword">return</span>
}</pre>
</div></div>
<p> </p></li>
</ol>


<p> </p>

<p> </p>

<p>autoTriggered() is the function I added that returns false if cause description contains "Started by...". All this does is check if it's a head job and if so it triggers the corresponding merge job and exits. </p>

<p>It feels a little too simple and too good to be true, but at first glance it seems like this gives best of both worlds of auto triggering merge jobs only on changes to head and avoiding the buildStorm. <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p> </p>
</pre>
</details>

</ul>
</details>

<p>I wanted to add here a solution I found that feels like a nice workaround at least for my needs:</p>


<ol>
	<li>In GitHub Org options check to build both merged and unmerged PRs so you have a bunch of PR jobs ending in -merge and -head.</li>
	<li>In the auto branch triggering field add PR-\d+-head so you auto build all head jobs but not all merge jobs. This will avoid the build storm that causes so many problems.</li>
	<li>From inside the Jenkinsfile add something like this:

<pre>
<span class="code-comment">// have head job trigger merge job...
</span><span class="code-keyword">if</span> (autoTriggered() && env.BRANCH_NAME.endsWith(<span class="code-quote">'head'</span>)) {
    def headJobName = <span class="code-quote">"${env.BRANCH_NAME}"</span>
    def mergeJobName = (headJobName.substring(0, headJobName.length() - 4)).concat(<span class="code-quote">'merge'</span>) <span class="code-comment">// replace head with merge
</span>    build job: mergeJobName, wait: <span class="code-keyword">false</span>
    <span class="code-keyword">return</span>
}</pre>
<p> </p></li>
</ol>


<p> </p>

<p> </p>

<p>autoTriggered() is the function I added that returns false if cause description contains "Started by...". All this does is check if it's a head job and if so it triggers the corresponding merge job and exits. </p>

<p>It feels a little too simple and too good to be true, but at first glance it seems like this gives best of both worlds of auto triggering merge jobs only on changes to head and avoiding the buildStorm. <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p> </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/seanf.png" title="seanf's avatar" /> seanf:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37491?focusedId=295506&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-295506">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<blockquote><p>With GitHub Branch Source 2.0.0-beta-1 (available from the <a href="https://jenkins.io/blog/2013/09/23/experimental-plugins-update-center/" class="external-link" target="_blank" rel="nofollow noopener">experimental update center</a> now or 2.0.0 (available in early January 2017)</p>

<p>When you push a change to the base branch, it should trigger only the base branch. The PR-merge branches will get rebuilt when the indexing kicks in as indexing will detect that the merge commit is now different, but that should only happen once a day/week (depending on how often you configure indexing) so should be much less of an issue</p></blockquote>
<p>Is it just me, or does that change cause a build storm at 9:00 every morning? (Assuming there's been a change in <b>master</b> branch.)</p>

<p>Is there any way of controlling the <em>time</em> of the daily indexing?</p>
</pre>
</details>

</ul>
</details>

<blockquote><p>With GitHub Branch Source 2.0.0-beta-1 (available from the <a href="https://jenkins.io/blog/2013/09/23/experimental-plugins-update-center/" class="external-link" target="_blank" rel="nofollow noopener">experimental update center</a> now or 2.0.0 (available in early January 2017)</p>

<p>When you push a change to the base branch, it should trigger only the base branch. The PR-merge branches will get rebuilt when the indexing kicks in as indexing will detect that the merge commit is now different, but that should only happen once a day/week (depending on how often you configure indexing) so should be much less of an issue</p></blockquote>
<p>Is it just me, or does that change cause a build storm at 9:00 every morning? (Assuming there's been a change in <b>master</b> branch.)</p>

<p>Is there any way of controlling the <em>time</em> of the daily indexing?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>spencermalone:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37491?focusedId=295654&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-295654">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I assumed this was based on the last time indexing happened + 24 hours. We normally get the storm in the AM as well, but I ran an index late last night so that I can see if it indexes in the evening today. If not, it's definitely supposed to be "Build if this hasn't built in the past day", so you could arguably setup another job or cron job to force an indexing at whatever time you want. Hopefully it's just doing "last indexing + 24 hours" as the behavior though.</p>

<p>To clarify, when I say indexing, I mean running a "Scan Repository Now" from the repo's folder page.</p>
</pre>
</details>

</ul>
</details>

<p>I assumed this was based on the last time indexing happened + 24 hours. We normally get the storm in the AM as well, but I ran an index late last night so that I can see if it indexes in the evening today. If not, it's definitely supposed to be "Build if this hasn't built in the past day", so you could arguably setup another job or cron job to force an indexing at whatever time you want. Hopefully it's just doing "last indexing + 24 hours" as the behavior though.</p>

<p>To clarify, when I say indexing, I mean running a "Scan Repository Now" from the repo's folder page.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/mikemcquaid.png" title="mikemcquaid's avatar" /> mikemcquaid:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37491?focusedId=295932&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-295932">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>The Multibranch pipeline's "Suppress automatic SCM triggering" for the `master` branch seems to stop this happening every time `master` changes. It still seems to happen whenever the repository is reindexed, however. Given there doesn't seem to be a current workaround I think we're going to have to go back to the GitHub Pull Request Builder plugin (which has its own issues). Note that the "Suppress automatic SCM triggering" doesn't seem to be available for GitHub Organisation jobs for no apparent reason.</p>

<p>As for our use-case: I'm a Homebrew (macOS package manager) maintainer and our builds are run on limited hardware and take anywhere from 5 minutes to 5 hours to build. Because of how our CI/testing functions and the number of packages we have 99.999% of the time `master` changes have no effect on a given PR (and when they do: we're on the ball enough to notice that). The way things are at the moment with ~70 open PRs our CI infrastructure simply can't come close to keeping up. Every time that all PRs are requeued it would take several days to clear the queue (at which point we've been unable to merge any other PRs for several days). I'm happy to talk more about our use-case if that will help.</p>

<p>My suggestions would be to ensure "Suppress automatic SCM triggering" also affects indexing and is available on GitHub Organisation jobs in short-term and in the medium term provide configuration to disable the master rebuild behave altogether. Opinionated design is great when your users don't have diverse workflows but that doesn't really apply in this case.</p>

<p>Regardless, thanks to everyone working on this for your work on Jenkins.</p>
</pre>
</details>

</ul>
</details>

<p>The Multibranch pipeline's "Suppress automatic SCM triggering" for the `master` branch seems to stop this happening every time `master` changes. It still seems to happen whenever the repository is reindexed, however. Given there doesn't seem to be a current workaround I think we're going to have to go back to the GitHub Pull Request Builder plugin (which has its own issues). Note that the "Suppress automatic SCM triggering" doesn't seem to be available for GitHub Organisation jobs for no apparent reason.</p>

<p>As for our use-case: I'm a Homebrew (macOS package manager) maintainer and our builds are run on limited hardware and take anywhere from 5 minutes to 5 hours to build. Because of how our CI/testing functions and the number of packages we have 99.999% of the time `master` changes have no effect on a given PR (and when they do: we're on the ball enough to notice that). The way things are at the moment with ~70 open PRs our CI infrastructure simply can't come close to keeping up. Every time that all PRs are requeued it would take several days to clear the queue (at which point we've been unable to merge any other PRs for several days). I'm happy to talk more about our use-case if that will help.</p>

<p>My suggestions would be to ensure "Suppress automatic SCM triggering" also affects indexing and is available on GitHub Organisation jobs in short-term and in the medium term provide configuration to disable the master rebuild behave altogether. Opinionated design is great when your users don't have diverse workflows but that doesn't really apply in this case.</p>

<p>Regardless, thanks to everyone working on this for your work on Jenkins.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>spencermalone:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37491?focusedId=310590&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-310590">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Opened a PR here: <a href="https://github.com/jenkinsci/branch-api-plugin/pull/109/files" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/branch-api-plugin/pull/109/files</a></p>

<p>So there's a property that is supposed to disable repository scan / branch indexing triggered builds, but it currently also disables all builds caused by events (such as webhooks, etc.) so it's not super useful for much right now. That PR should fix it, so that we can disable org scans triggering builds by going...</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>properties([overrideIndexTriggers(false)])</pre>
</div></div>
<p>in a Jenkinsfile.</p>

<p> </p>

<p>In my mind, this would satisfy the ticket, does anyone have other thoughts?</p>
</pre>
</details>

</ul>
</details>

<p>Opened a PR here: <a href="https://github.com/jenkinsci/branch-api-plugin/pull/109/files" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/branch-api-plugin/pull/109/files</a></p>

<p>So there's a property that is supposed to disable repository scan / branch indexing triggered builds, but it currently also disables all builds caused by events (such as webhooks, etc.) so it's not super useful for much right now. That PR should fix it, so that we can disable org scans triggering builds by going...</p>


```
properties([overrideIndexTriggers(false)])
```
<p>in a Jenkinsfile.</p>

<p> </p>

<p>In my mind, this would satisfy the ticket, does anyone have other thoughts?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/stephenconnolly.png" title="stephenconnolly's avatar" /> stephenconnolly:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37491?focusedId=310597&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-310597">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>&gt; but it currently also disables all builds caused by events (such as webhooks, etc.) so it's not super useful for much right now. </p>

<p>Actually there is a cohort of people who consider this behaviour as exactly the behaviour that they want... they want the branch projects created but they do not want those branch projects to ever have an automatic build.</p>

<p>My intent is to remove that property (should never have been added) and replace it with the concept of interesting heads/revisions and non-interesting heads/revisions (<a href="https://issues.jenkins.io/browse/JENKINS-45502" title="Add concept of &quot;interesting&quot; to SCMHead and SCMRevision" class="issue-link" data-issue-key="JENKINS-45502">JENKINS-45502</a>) so for that reason I will most likely reject PRs on this functionality against the branch-api plugin (as these kinds of tweaks should not be taking place in the branch-api plugin) Feel free to create an extension plugin that provides the behaviour you want. I will merge PRs that enable extension plugins to do this kind of thing, with the proviso that I believe <a href="https://issues.jenkins.io/browse/JENKINS-45502" title="Add concept of &quot;interesting&quot; to SCMHead and SCMRevision" class="issue-link" data-issue-key="JENKINS-45502">JENKINS-45502</a> is the correct direction</p>

</pre>
</details>

</ul>
</details>

<p>> but it currently also disables all builds caused by events (such as webhooks, etc.) so it's not super useful for much right now. </p>

<p>Actually there is a cohort of people who consider this behaviour as exactly the behaviour that they want... they want the branch projects created but they do not want those branch projects to ever have an automatic build.</p>

<p>My intent is to remove that property (should never have been added) and replace it with the concept of interesting heads/revisions and non-interesting heads/revisions (<a href="https://issues.jenkins.io/browse/JENKINS-45502" title="Add concept of "interesting" to SCMHead and SCMRevision" class="issue-link" data-issue-key="JENKINS-45502">JENKINS-45502</a>) so for that reason I will most likely reject PRs on this functionality against the branch-api plugin (as these kinds of tweaks should not be taking place in the branch-api plugin) Feel free to create an extension plugin that provides the behaviour you want. I will merge PRs that enable extension plugins to do this kind of thing, with the proviso that I believe <a href="https://issues.jenkins.io/browse/JENKINS-45502" title="Add concept of "interesting" to SCMHead and SCMRevision" class="issue-link" data-issue-key="JENKINS-45502">JENKINS-45502</a> is the correct direction</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>spencermalone:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37491?focusedId=310605&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-310605">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>But those that want to disable all options could just use both </p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>properties([overrideIndexTriggers(false), overrideEventTriggers(false)])</pre>
</div></div>
<p>My problem is that currently, those of us that want organization scans to not trigger builds, but still want to run them for the other things they do (such as old branch job cleanup, or finding "lost" PRs) are kind of left without a way to do what we need. This patch lets people choose their default behavior on a per-branch basis.</p>

<p>I agree that the original property seems like a mistake when combined with everything else.</p>

<p>How far out is the interesting head/revision idea? I understand the desire to push forward on a big revision like that, but this has been a pretty big blocker for us for a while. The patch is relatively simple, and it seems like the interesting head/revision stuff is way off on the horizon. Currently this is a key issue that we're feeling every day. </p>

<p> </p>

<p>EDIT: Sorry, after rereading, I see more of the emphasis on the creating extension points. I'll look into that, but this is my first contribution to jenkins plugins, so it'll take me some time to get used to how extensions work.</p>
</pre>
</details>

</ul>
</details>

<p>But those that want to disable all options could just use both </p>


```
properties([overrideIndexTriggers(false), overrideEventTriggers(false)])
```
<p>My problem is that currently, those of us that want organization scans to not trigger builds, but still want to run them for the other things they do (such as old branch job cleanup, or finding "lost" PRs) are kind of left without a way to do what we need. This patch lets people choose their default behavior on a per-branch basis.</p>

<p>I agree that the original property seems like a mistake when combined with everything else.</p>

<p>How far out is the interesting head/revision idea? I understand the desire to push forward on a big revision like that, but this has been a pretty big blocker for us for a while. The patch is relatively simple, and it seems like the interesting head/revision stuff is way off on the horizon. Currently this is a key issue that we're feeling every day. </p>

<p> </p>

<p>EDIT: Sorry, after rereading, I see more of the emphasis on the creating extension points. I'll look into that, but this is my first contribution to jenkins plugins, so it'll take me some time to get used to how extensions work.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37491?focusedId=311144&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-311144">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>This is something that should be solved in <tt>github-branch-source</tt>, not <tt>branch-api</tt>, though a common extension in <tt>scm-api</tt> might be useful so the same UI can be reused for BitBucket etc.</p>
</pre>
</details>

</ul>
</details>

<p>This is something that should be solved in <tt>github-branch-source</tt>, not <tt>branch-api</tt>, though a common extension in <tt>scm-api</tt> might be useful so the same UI can be reused for BitBucket etc.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>acejam:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37491?focusedId=332208&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-332208">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I'm also now seeing build storms from this issue.</p>

<p>One can turn off "Periodically if not otherwise run" to prevent the storms, but then you're left with orphaned jobs being left behind when a branch is deleted. Therefore, if one tries to use the branch source plugin, they either have to deal with build storms, or with orphaned jobs.</p>

<p>I don't think I saw this issue back when the core functionality existed in the GitHub Organization Folders plugin. Perhaps this is a regression?</p>
</pre>
</details>

</ul>
</details>

<p>I'm also now seeing build storms from this issue.</p>

<p>One can turn off "Periodically if not otherwise run" to prevent the storms, but then you're left with orphaned jobs being left behind when a branch is deleted. Therefore, if one tries to use the branch source plugin, they either have to deal with build storms, or with orphaned jobs.</p>

<p>I don't think I saw this issue back when the core functionality existed in the GitHub Organization Folders plugin. Perhaps this is a regression?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/stephenconnolly.png" title="stephenconnolly's avatar" /> stephenconnolly:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37491?focusedId=332268&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-332268">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>For the record, if you want to avoid triggering rebuilds when just the target branch of the PR has changed then you want <a href="https://github.com/jenkinsci/basic-branch-build-strategies-plugin/blob/master/docs/user.adoc#change-requests" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/basic-branch-build-strategies-plugin/blob/master/docs/user.adoc#change-requests</a> </p>

<p> </p>

<p>IMHO this problem is closed by the feature "Ignore rebuilding merge branches when only the target branch changed" in the Basic Branch Build Strategy plugin's "Change Requests" strategy</p>
</pre>
</details>

</ul>
</details>

<p>For the record, if you want to avoid triggering rebuilds when just the target branch of the PR has changed then you want <a href="https://github.com/jenkinsci/basic-branch-build-strategies-plugin/blob/master/docs/user.adoc#change-requests" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/basic-branch-build-strategies-plugin/blob/master/docs/user.adoc#change-requests</a> </p>

<p> </p>

<p>IMHO this problem is closed by the feature "Ignore rebuilding merge branches when only the target branch changed" in the Basic Branch Build Strategy plugin's "Change Requests" strategy</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/stephenconnolly.png" title="stephenconnolly's avatar" /> stephenconnolly:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37491?focusedId=332269&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-332269">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Resolution:</p>
<ul>
	<li>Install Basic Branch Build Strategies plugin and add the "Change Requests" with "Ignore rebuilding merge branches when only the target branch changed" selected to your build strategies</li>
</ul>

</pre>
</details>

</ul>
</details>

<p>Resolution:</p>
<ul>
	<li>Install Basic Branch Build Strategies plugin and add the "Change Requests" with "Ignore rebuilding merge branches when only the target branch changed" selected to your build strategies</li>
</ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37491?focusedId=336023&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-336023">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>For safety, you may want to then configure GitHub to refuse to let PRs be merged unless they are up to date with the base branch—IOW to only permit fast-forward merges from the GUI.</p>
</pre>
</details>

</ul>
</details>

<p>For safety, you may want to then configure GitHub to refuse to let PRs be merged unless they are up to date with the base branch—IOW to only permit fast-forward merges from the GUI.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>acejam:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37491?focusedId=336028&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-336028">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I can confirm that setting up the Basic Branch Build Strategies plugin and selecting <em>Regular Branches</em> and <em>Change Requests</em> (with <em>Ignore rebuilding merge branches</em> checked) indeed gives us exactly what we want. Thanks for the tip!</p>
</pre>
</details>

</ul>
</details>

<p>I can confirm that setting up the Basic Branch Build Strategies plugin and selecting <em>Regular Branches</em> and <em>Change Requests</em> (with <em>Ignore rebuilding merge branches</em> checked) indeed gives us exactly what we want. Thanks for the tip!</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>aldochristiaan:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37491?focusedId=352718&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-352718">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=stephenconnolly" class="user-hover" rel="stephenconnolly">stephenconnolly</a> hey, i've installed the plugins, but i can't find the options on my multiple pipeline job .. Could you help me please ?</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=stephenconnolly" class="user-hover" rel="stephenconnolly">stephenconnolly</a> hey, i've installed the plugins, but i can't find the options on my multiple pipeline job .. Could you help me please ?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>brianjmurrell:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-37491?focusedId=359043&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-359043">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=aldochristiaan" class="user-hover" rel="aldochristiaan">aldochristiaan</a> By <em>multiple pipeline job</em> do you mean GitHub Organisation jobs?  Because I agree, the proposed solution here doesn't seem to apply, AFAICT, to the GitHub Source Branch Plugin multibranch-pipeline job.</p>

<p>So those users still need a solution to these build storms that just take down your Jenkins machine every time a base branch (i.e. master) is updated.  I guess we need <a href="https://issues.jenkins-ci.org/browse/JENKINS-55822" class="external-link" target="_blank" rel="nofollow noopener">another ticket</a>.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=aldochristiaan" class="user-hover" rel="aldochristiaan">aldochristiaan</a> By <em>multiple pipeline job</em> do you mean GitHub Organisation jobs?  Because I agree, the proposed solution here doesn't seem to apply, AFAICT, to the GitHub Source Branch Plugin multibranch-pipeline job.</p>

<p>So those users still need a solution to these build storms that just take down your Jenkins machine every time a base branch (i.e. master) is updated.  I guess we need <a href="https://issues.jenkins-ci.org/browse/JENKINS-55822" class="external-link" target="_blank" rel="nofollow noopener">another ticket</a>.</p>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `is_duplicated_by` from Jira: <a href="https://github.com/jenkinsci/github-branch-source-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-55822%22">JENKINS-55822</a>]</i>
